controllers/preCachingConfig.go

```diff
@@ -0,0 +1,76 @@
+package controllers
+
+import (
+	"context"
+	"reflect"
+
+	ranv1alpha1 "github.com/openshift-kni/cluster-group-upgrades-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// getPreCachingConfig: retrieves pre-caching configuration CR
+func (r *ClusterGroupUpgradeReconciler) getPreCachingConfig(
+	ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) (
+	*ranv1alpha1.PreCachingConfig, error) {
+
+	preCachingConfig := new(ranv1alpha1.PreCachingConfig)
+	preCachingConfigRef := clusterGroupUpgrade.Spec.PreCachingConfigRef
+
+	preCachingConfigName := preCachingConfigRef.Name
+	preCachingConfigNamespace := preCachingConfigRef.Namespace
+	// If namespace is not specified, assume the CGU namespace
+	if preCachingConfigNamespace == "" {
+		preCachingConfigNamespace = clusterGroupUpgrade.Namespace
+	}
+
+	if preCachingConfigName == "" {
+		r.Log.Info("getPreCachingConfigSpec: no preCachingConfig CR specified")
+		return preCachingConfig, nil
+	}
+
+	err := r.Get(ctx, types.NamespacedName{Namespace: preCachingConfigNamespace,
+		Name: preCachingConfigName}, preCachingConfig)
+
+	if err != nil {
+		return preCachingConfig, err
```

Comment:
Not sure if the err print will look good in the condition. Might be a good idea to wrap the error with additional text. You need to manually test e2e at least once if you don't want to add new step and it should look like this:

```
  - message: "Precaching spec is incomplete: inconsistent precaching configuration: olm index provided, but no packages"
    reason: PrecacheSpecIncomplete
    status: "False"
    type: PrecacheSpecValid
```

Reply:
Done! A new TestStep has been added to `tests/kuttl/tests/pre-caching-complete-with-configs` to verify this behaviour.